src/decoder/cuda-decoder-utils.h

```diff
@@ -0,0 +1,431 @@
+#define MEMADVISE // used after Pascal, details in: 
+// http:// mug.mvapich.cse.ohio-state.edu/static/media/mug/presentations/2016/MUG16_GPU_tutorial_V5.pdf
+#endif
+
+#define DIV_ROUND_UP(a,b) ((a+b-1)/b)
+
+// decode & encode function of tok address, used in host & device
+// pack frame & index into uint64 as the address of a tok
+#define ENCODE_TOK_IDX_PAIR(frame,idx) (((uint64)(frame)<<32)+(idx))
+// get frame & index in the per-frame vector of a tok address packed in uint64
+#define DECODE_TOK_IDX_PAIR(frame,idx,v) { \
+    frame=(((uint64)v)>>32); \
+    idx=(((uint64)v)&(((uint64)1<<32)-1)); \
+}
+
+// host function definition
+
+// get current GPU memory usage
+void GetFreeMemoryStat(const char *prefix);
+
+// a combination of cudaMallocManaged & cudaMemAdvise
+void CudaMallocManagedPreferredDevice(void** devPtr, size_t size);
+
+// inline host device function definition
+
+union float_uint {
+  float f;
+  uint32 u;
+};
+inline HOST DEVICE uint32 float_as_uint(BaseFloat val) {
+  return ((float_uint*)&val)->u;
+}
+inline HOST DEVICE BaseFloat uint_as_float(uint32 val) {
+  return ((float_uint*)&val)->f;
+}
+
+// In atomic based token recombination, we pack the
+// cost and the arc index into an uint64 to represent the token
+// before recombination, with the former one in the higher bits
+// for comparison purpose.
+// for speedup purpose, make them inline (5% 0.165->0.158)
+inline HOST DEVICE uint64 pack_cost_idx_into_uint64(BaseFloat cost, int32 idx) {
+  uint32 i_cost = float_as_uint(cost);
+  if (i_cost & 0x80000000)
+    i_cost = i_cost ^ 0xFFFFFFFF;
+  else
+    i_cost = i_cost ^ 0x80000000;
+  return (uint64)i_cost << 32 | idx;
+}
+
+// Unpacks a cost
+inline HOST DEVICE BaseFloat unpack_cost_from_uint64(uint64 packed) {
+  uint32 i_cost = packed >> 32;
+  if (i_cost & 0x80000000)
+    i_cost = i_cost ^ 0x80000000;
+  else
+    i_cost = i_cost ^ 0xFFFFFFFF;
+  return uint_as_float(i_cost);
+}
+
+// Unpacks a idx for tracing the data
+inline HOST DEVICE int32 unpack_idx_from_uint64(uint64 packed) {
+  return packed & 0x7FFFFFFF;
+}
+
+// inline device function definition
+#ifdef __CUDACC__
+
+// another choice is pre-store the result of this binsearch, e.g. do something like:
+// for (int j=0; j < params->d_degrees[idx]; j++)
+//  params->d_lowerbound[j + params->d_degrees_scan[idx]] = idx;
+// however this for loop costs 9% more time, while the binsearch costs 6% time
+// thus the fastest way is doing binsearch without pre-store
+inline DEVICE int binsearch_maxle(const int *vec, const int val, int low,
+                                  int high) {
+  while (true) {
+    if (low == high)
+      return low; //we know it exists
+    if ((low + 1) == high)
+      return (vec[high] <= val) ? high : low;
+
+    int mid = low + (high - low) / 2;
+
+    if (vec[mid] > val)
+      high = mid - 1;
+    else
+      low = mid;
+  }
+}
+
+
+// fast load 16 bits using CUDA ASM
+inline  DEVICE void fast_load16(void *a, const void *b) {
```

Comment:
fast_load16 appears unused.

Reply:
how about leaving it as a pair with fast_store16